stream_alert/alert_processor/outputs/pagerduty.py

```diff
@@ -610,70 +623,131 @@ def run(self, alert, descriptor):
+        incident = self._update_base_incident(event, alert, publication, rule_context)
+        if not incident:
             LOGGER.error(
-                '[%s] Failed to merge alert [%s] into [%s]',
-                self._output.__service__,
-                event.get('dedup_key'),
-                incident_id
+                '[%s] Failed to update container incident for event',
+                self._output.__service__
             )
-            stable = False
+            return False
 
-        if merged_incident:
-            note = self._add_incident_note(incident, publication, rule_context)
-            if not note:
-                LOGGER.error(
-                    '[%s] Failed to add note to incident (%s)',
-                    self._output.__service__,
-                    incident_id
+        incident_id = incident.get('id', False)
+        if not incident_id:
+            LOGGER.error(
+                '[%s] Incident is missing "id"??',
+                self._output.__service__
+            )
+            return False
+
+        # At this point, both the incident and the relevant alert event have been successfully
+        # created.
+        #
+        # All of the code above this line is considered idempotent and can be called repeatedly
+        # without adverse side effects. Code BELOW this line is neither atomic nor idempotent, so
+        # we will not retry if any of the below code fails. Instead, we log an error and make a
+        # best-effort attempt to attach an error note to the PagerDuty incident, signalling that
+        # it was not setup properly.
+        #
+        # In the fateful event the alert gets stuck ANYWAY, the easiest solution is to destroy the
+        # associated record on the DynamoDB table.
+        errors = []
+
+        # Add responder requests
+        responders = rule_context.get('responders', [])
+        if responders and not isinstance(responders, list):
+            responders = [responders]
+
+        if responders:
+            # The message shows up in the email
+            default_message = 'An incident was reported that requires your attention.'
+            responder_message = rule_context.get('responder_message', default_message)
+
+            for responder_email in responders:
+                result = self._add_incident_response_request(
+                    incident_id,
+                    responder_email,
+                    responder_message
                 )
-                stable = False
+                if not result:
+                    error = '[{}] Failed to request a responder ({}) on incident ({})'.format(
+                        self._output.__service__,
+                        responder_email,
+                        incident_id
+                    )
+                    LOGGER.error(error)
+                    errors.append(error)
+
+        # Add a note to the incident
+        note = self._add_incident_note(incident_id, publication, rule_context)
```

Comment:
where do we add this `note` to?

Reply:
This note is a small string of text that is added to the PagerDuty incident. It shows up on their UI